presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -358,7 +361,52 @@ else if (firstDependencies.stream().allMatch(right::canResolve) && secondDepende
         return new RelationPlan(root, analysis.getScope(node), outputSymbols);
     }
 
-    private static boolean isEqualComparisonExpression(Expression conjunct)
+    private Optional<Unnest> getUnnest(Relation relation)
+    {
+        if (relation instanceof AliasedRelation) {
+            return getUnnest(((AliasedRelation) relation).getRelation());
+        }
+        if (relation instanceof Unnest) {
+            return Optional.of((Unnest) relation);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<Lateral> getLateral(Relation relation)
+    {
+        if (relation instanceof AliasedRelation) {
+            return getLateral(((AliasedRelation) relation).getRelation());
+        }
+        if (relation instanceof Lateral) {
+            return Optional.of((Lateral) relation);
+        }
+        return Optional.empty();
+    }
+
+    private RelationPlan planLateralJoin(Node joinNode, RelationPlan leftPlan, Lateral lateral)
+    {
+        RelationPlan rightPlan = process(lateral.getQuery(), null);
+        PlanBuilder leftPlanBuilder = initializePlanBuilder(leftPlan);
+        PlanBuilder planBuilder = subqueryPlanner.appendApplyNode(
```

Comment:
Additionally, `PruneUnreferencedOutputs` removes `ApplyNodes` when it's subquery projection symbols are not used. If one models laterals as `ApplyNode` this is not longer valid optimization since subquery cardinality might also be relevant (e.g: for `count(*)`). Therefore, one cannot remove `ApplyNodes` even if it's subquery symbols are not used.